Added Monotone Chain Algorithm and implementations to Convex Hull module (plat) (#1928)
cpinitiative/usaco-guide: content/5_Plat/Convex_Hull.mdx

### Solution

> With the Monotone Chain algorithm, we start by sorting the given $n$ points in ascending order with respect to their $x$ coordinates. If two points have the same $x$ value, then we will look at the $y$ coordinate. 

Reviewer: ```suggestion
With the Monotone Chain algorithm, we start by sorting the given $n$ points in ascending order with respect to their $x$ coordinates. If two points have the same $x$ coordinate, then we will look at the $y$ coordinate. 
```

> Next, we will calculate the convex hull in two parts - the upper and the lower hull. Firstly, we observe that the starting and ending points of both upper and lower hulls are the same. They are the points with the lowest and highest $x$ value, $P_0$ and $P_{n-1}$. We primarily add $P_0$ and $P_1$ to the hull. Note that $P_1$ doesn't necessarily have to be a point of the convex hull at the end. Then, starting from $P_2$, we iterate through the sorted points and add them to the hull. Let's denote the current point being inspected as $P_{k}$ and the last point in the hull as $P_{i}$. While adding new points, we want to make sure that there is no right turn among all segments of the hull, just like in the Graham Algorithm discussed above. To achieve this, the segment $P_{i-1}P_i$ should always be on the left side of the segment $P_{i-1}P_k$. This can be calculated by a cross product trick: 

Reviewer: ```suggestion
Next, we will calculate the convex hull in two parts - the upper and the lower hull. Firstly, we observe that the starting and ending points of both upper and lower hulls are the same. They are the points with the lowest and highest $x$ value respectively, $P_0$ and $P_{n-1}$. We start by adding $P_0$ and $P_1$ to the hull. (Note that $P_1$ doesn't necessarily have to be on the convex hull at the end). Then, starting from $P_2$, we iterate through the sorted points and add them to the hull. Let's denote the current point being added as $P_{k}$ and the last point still on the hull as $P_{i}$. When adding new points, we want to make sure that there is no right turn among all segments of the hull, just like in the Graham Scan algorithm discussed above. To achieve this, the segment $P_{i-1}P_i$ should always be on the left side of the segment $P_{i-1}P_k$. This can be calculated by using a cross-product: 
```

- If $(P_i - P_{i-1}) \times (P_k - P_{i-1}) < 0$, the point $P_i$ lies on the left side of the segment $P_{i-1}P_k$. In this case, we have to remove point $P_i$ from the hull and repeat this check. 

> - If $(P_i - P_{i-1}) \times (P_k - P_{i-1}) = 0$, the point $P_i$ lies on the segment $P_{i-1}P_k$. Whether to include multiple collinear points depends on the question, but to the given question above, we will remove the point $P_i$ as well and repeat the check. 

Reviewer: ```suggestion
- If $(P_i - P_{i-1}) \times (P_k - P_{i-1}) = 0$, the point $P_i$ lies on the segment $P_{i-1}P_k$. Whether to include multiple collinear points depends on the question, but for the given question above, we will remove the point $P_i$ as well and repeat the check. 
```

> - Otherwise, the point $P_i$ lies on the right side of the segment $P_{i-1}P_k$. So we are good to inspect the next point $P_{k+1}$ from the given point list. 

Reviewer: ```suggestion
- Otherwise, the point $P_i$ lies on the right side of the segment $P_{i-1}P_k$. In this case, we can process the next point $P_{k+1}$ from the given point list. 
```

> After all points being examined, we have calculated the lower hull and will begin to find the upper hull in the same manner. This time, we add point $P_{n-2}$ to the hull and iterate from the end of the points, $P_{n-3}$, to the starting point $P_0$. The Point $P_{n-2}$ doesn't have to be the convex hull and could be removed if it causes a right turn. 

Reviewer: ```suggestion
After all the points have been processed, we have found the lower hull and will begin to find the upper hull in the same manner. This time, we add point $P_{n-2}$ to the hull and iterate from the end of the points, $P_{n-3}$, to the starting point $P_0$. (The point $P_{n-2}$ also doesn't necessarily have to be the convex hull and could be removed if it causes a right turn). 
```
Also do you mean we've found the *upper* hull? (Judging from your cross product stuff)

Author: opps, the lower hull is correct, but I messed up with this: 
To achieve this, the segment $P_{i-1}P_i$ should always be on the <del>left</del> **right** side of the segment $P_{i-1}P_k$

gonna fix it

> At the end, we have traversed the points in the counterclockwise order. To do this in the clockwise order, one only has to change the condition for $(P_i - P_{i-1}) \times (P_k - P_{i-1})$ from more than 0 to less than 0. In this case, the upper hull will be found first and then the lower hull. 

Reviewer: > At the end, we have traversed the points in the counterclockwise order.

We have? Do you rather mean we've found the points on the convex hull in counterclockwise order?

Author: ya we got the points of the hull in counterclockwise order, not the traversal. Im gonna fix it.

This algorithm takes $\mathcal{O}(n \log n)$ time to sort the points and $\mathcal{O}(n)$ time to calculate the hull, giving a final time complexity of $\mathcal{O}(n \log n)$. 

<Spoiler title="Worked Example">

> Let's consider the same example from the Graham Algorithm. 

Reviewer: ```suggestion
Let's consider the same example from the Graham Scan algorithm. 
```

3. Find the upper hull: 

   We first add the point $(5,3)$ to the end of the list and process the point $(3,5)$. The cross product for $((6,2) - (5,3)) \times ((6,2) - (3,5))$ is zero, meaning these three points are collinear. Therefore, we remove point $(5,3)$ from the list. As there is no enough points left in the upper hull, we add point $(3,5)$ to the list. The hull contains $[(1,2), (2,0), (6,2), (3,5)]$. Note the points $[(1,2), (2,0), (6,2)]$ belong to the lower hull and the point $(6,2)$ as the end point also belongs to the upper hull.  

	 Then we process the next point $(2,3)$. The segments $(6,2)(3,5)$ and $(3,5)(2,3)$ form a left turn, so we can add point $(2,3)$ to our hull. The list contains $[(1,2), (2,0), (6,2), (3,5), (2,3)]$. 

	 The next point is $(2,0)$. The segments $(3,5)(2,3)$ and $(2,3)(2,0)$ form a left turn, so we can add point $(2,3)$ to our hull. The list contains $[(1,2), (2,0), (6,2), (3,5), (2,3), (2,0)]$.

> 	The next point is $(1,2)$. There is a right turn from $(2,3)(2,0)$ to $(2,0)(1,2)$, so we remove point $(2,0)$ from the list. The list contains $[(1,2), (2,0), (6,2), (3,5), (2,3)]$. We test again for segments $(3,5)(2,3)$ and $(2,3)(1,2)$. There is a right turn, so we remove point $(2,3)$. The list contains $[(1,2), (2,0), (6,2), (3,5)]$. We once again test segments $(6,2)(3,5)$ and $(3,5)(1,2)$. There is a left turn, so we add point $(1,2)$ to the list. We have reached the beginning of our list so we can terminate here. As $(1,2)$ was added while calculating the lower hull, we then remove the last element, i.e. the added point $(1,2)$, from the list. Now, we have the upper hull calculated as well. The final result is $[(1,2), (2,0), (6,2), (3,5)]$. These are the vertices of our convex hull in counterclockwise order. 

Reviewer: I see that there are line breaks, but the page looks really hard to read:
![image](https://user-images.githubusercontent.com/31325319/127863395-09fb12f1-65a3-4b85-8bf2-844b37795ef9.png)

@thecodingwizard is this a website bug?

Reviewer 2: Highly probable oops. If everything else is fine we can just merge this PR and then open a new GitHub issue for this :P
